LCB Package Signing and Module Protection

runrevmark opened the pull request:
WIP

runrevmark committed:
[[ Engine ]] Command-line 'internal' access.

The development engine can now run 'internal' commands by running '<engine> -internal …' from the command-line.

In order for this to work, the engine must be licensed on the system (i.e. it must have been run on the Desktop and license details entered).

The internal actions are run in 'noui' mode and are limited to the ones defined in MCU_internal_action.

runrevmark committed:
[[ Engine ]] Trust / protection primitives.

Added implementation of internal commands 'builder validate' and 'builder protect' for community.

runrevmark committed:
[[ Foundation ]] MCMemoryOutputStreamFinishAsDataRef API.

runrevmark committed:
[[ Engine ]] Update project files.

Added 'deploy_builder.cpp' to Linux and Windows project files.

runrevmark committed:
[[ Engine ]] Implemented basic primitive of inserting a block of bytes into a zip comment field.

peter-b commented:
[Mozilla NSS SignTool](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Tools/signtool) is a good example of a very standard approach to storing signatures in a `.zip`/`.jar` file (i.e. `META-INF` subdirectory containing signatures).  It's used for signing Firefox extensions.

APK files are also signed by (you guessed it) putting signatures in the `META-INF` subdirectory.

Do we really want to go against the industry standard?

peter-b commented:
Content signing in OpenDocument containers (which are ZIP archives) is implemented by storing signatures in the `META-INF` subdirectory.

runrevmark commented:
Certainly the META-INF standard approaches are more general (it appears that OpenDoc uses XML files, whereas JAR-related ones use .sf/.mf files) - they work using per-file signing within the archive, which is obviously more fine-grained and means you can modify the zip after the event as long as you don't modify any of the explicitly checksummed files.

Android actually supports whole-file signed zips in addition to the META-INF approach - it does exactly the same thing we do here - inserting the signature in the terminal comment. (Android uses such things for OTA update packages - rather than APK packages).

So I'm slightly on the fence. The whole-file approach is certainly simpler and easier to implement; but then if per-file signing provides a tangible benefit then its worth considering. (To be fair implementation-wise - both rely on the same fundamental operation produce a signature for a sequence of bytes - the whole zip for one, the manifest.mf for per-file signing; the difference only comes in mutating the original zip).

peter-b commented:
> Android actually supports whole-file signed zips in addition to the META-INF approach - it does exactly the same thing we do here - inserting the signature in the terminal comment. (Android uses such things for OTA update packages - rather than APK packages).

Oh, cool! I completely missed that when I was looking into the way Android does stuff.  Can we steal their format?  It might be a good way to package stuff initially, even if we it means we have to give up on multi-signature packages to start off with.

On a slight side-note, what about adopting the OpenDocument technique of storing an uncompressed `mimetype` file as the first file in the zip archive?  It'll make it much easier for tools like `file` to reliably identify misnamed LiveCode packages (and to show them as LiveCode packages rather than zip archives).

runrevmark commented:
The android 'whole-file' approach is described as 'custom' (it's mentioned in a few overviews of the Android APK security model) but it seems to fit what we need with LCB packages at the moment - particularly from the point of view of verifying the person uploading the (whole) package is a person who can access their extensions account.

In regards to the per-file signing, this is perhaps something we can add subsequently as we improve and make the trust model more flexible. We'll also need to do 'something' with regards foreign code module signing as well, since that's required to run on Windows / Mac without dialogs popping up (and to ensure LiveCode app builders find it easy to be 'good' citizens).

I noticed the OpenDoc approach to interacting with 'file' and also thought it a good idea :)

peter-b commented:
Signed Linux kernel modules use an appended X.509 signature which seems to include all the information we care about. Might be possible to reuse that format.

runrevmark commented:
The X.509 format is extensible anyway - Authenticode (the MS model we have code for already) uses X.509 signatures, it just requires some specific extra fields to be recognised as a valid signature for use on Windows. OpenSSL happily manipulates such things relatively easily (once you get your head around the way the data structures work!).

runrevmark committed:
[[ Engine ]] Update to package signing.

The package signing primitive now uses a whole-archive format borrowed from Android APK OTA signing.

The container for the signature is written into commentless archives, or replaced in archives with an existing signature - the actual binary signature generation is still to do.

livecode-vulcan commented:
📝 Hi @runrevmark, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.

montegoulding commented:
Should this be resurrected or should we just use a third party command line tool?

runrevmark commented:
@montegoulding: I'm not too fussed either way. This code embeds a signature in the comment field of a zip archive - the same mechanism is used by Android's update zips. The advantage is that a signature can be added / removed without fettling with the contents of the zip - which might still be a useful model for resigning / signing things with more than one key.

Remember the issue with using third-party command line tools is that we have to maintain and compile them for all desktop platforms _and_ be able to distribute them with the IDE.

montegoulding commented:
I was thinking of just using jarsigner as it's already available for all the desktop platforms and required for one of the platforms we deploy on anyway
